src/modules/rlm_ldap/groups.c

```diff
@@ -837,254 +807,329 @@ unlang_action_t rlm_ldap_check_groupobj_dynamic(rlm_rcode_t *p_result, rlm_ldap_
+		/*
+		 *	To avoid re-assessing after each call out to do a DN -> name
+		 *	lookup, cache this.
+		 */
+		group_ctx->count = ldap_count_values_len(group_ctx->values);
 	}
 
-	entry = ldap_first_entry(query->ldap_conn->handle, query->result);
-	if (!entry) {
-		ldap_get_option(query->ldap_conn->handle, LDAP_OPT_RESULT_CODE, &ldap_errno);
-		REDEBUG("Failed retrieving entry: %s", ldap_err2string(ldap_errno));
+	/*
+	 *	Following a call out to do a DN -> name lookup, group_ctx->query will be
+	 *	populated - process the results.
+	 */
+	if (group_ctx->query) {
+		char		*buff;
+		struct berval	**values = NULL;
 
-		rcode = RLM_MODULE_FAIL;
+		switch (group_ctx->query->ret) {
+		case LDAP_RESULT_SUCCESS:
+			break;
 
-		goto finish;
-	}
+		case LDAP_RESULT_NO_RESULT:
+		case LDAP_RESULT_BAD_DN:
+			REDEBUG("Group DN \"%s\" did not resolve to an object", group_ctx->lookup_dn);
+			RETURN_MODULE_INVALID;
 
-	values = ldap_get_values_len(query->ldap_conn->handle, entry, inst->userobj_membership_attr);
-	if (!values) {
-		RDEBUG2("No group membership attribute(s) found in user object");
+		default:
+			RETURN_MODULE_FAIL;
+		}
 
-		goto finish;
+		entry = ldap_first_entry(group_ctx->query->ldap_conn->handle, group_ctx->query->result);
+		if (!entry) {
+			ldap_get_option(group_ctx->query->ldap_conn->handle, LDAP_OPT_RESULT_CODE, &ldap_errno);
+			REDEBUG("Failed retrieving entry: %s", ldap_err2string(ldap_errno));
+			RETURN_MODULE_INVALID;
+		}
+
+		values = ldap_get_values_len(group_ctx->query->ldap_conn->handle, entry, inst->groupobj_name_attr);
+		if (!values) {
+			REDEBUG("No %s attributes found in object", inst->groupobj_name_attr);
+			RETURN_MODULE_INVALID;
+		}
+
+		buff = talloc_bstrndup(group_ctx, values[0]->bv_val, values[0]->bv_len);
+		RDEBUG2("Group DN \"%s\" resolves to name \"%s\"", group_ctx->lookup_dn, buff);
+		ldap_value_free_len(values);
+
+		if (group_ctx->resolving_value) {
+			value_name = buff;
+		} else {
+			group_ctx->group_name = buff;
+		}
 	}
 
 	/*
-	 *	Loop over the list of groups the user is a member of,
-	 *	looking for a match.
+	 *	Loop over the list of groups the user is a member of, looking for a match.
 	 */
-	name_is_dn = fr_ldap_util_is_dn(check->vp_strvalue, check->vp_length);
-	count = ldap_count_values_len(values);
-	for (i = 0; i < count; i++) {
-		value_is_dn = fr_ldap_util_is_dn(values[i]->bv_val, values[i]->bv_len);
+	while (group_ctx->value_no < group_ctx->count) {
+		struct berval	*value = group_ctx->values[group_ctx->value_no];
+
+		/*
+		 *	We have come back from resolving a membership DN to its name,
+		 *	compare to the provided name.
+		 */
+		if (value_name && group_ctx->resolving_value) {
+			if (((talloc_array_length(value_name) - 1) == group->vb_length) &&
+			    (memcmp(group->vb_strvalue, value_name, group->vb_length) == 0)) {
+				RDEBUG2("User found in group \"%pV\". Comparison between membership: name "
+				       "(resolved from DN \"%s\"), check: name", group, group_ctx->lookup_dn);
```

Comment:
Should technically be binary safe when printing group_ctx->lookup_dn.  Use one of the `fr_box_strvalue` macros